agent/engine/docker_container_engine.go

```diff
@@ -218,38 +225,63 @@ func (dg *dockerGoClient) time() ttime.Time {
 }
 
 func (dg *dockerGoClient) PullImage(image string, authData *api.RegistryAuthenticationData) DockerContainerMetadata {
+	// TODO Switch to just using context.WithDeadline and get rid of this funky code
 	timeout := dg.time().After(pullImageTimeout)
+	ctx, cancel := context.WithCancel(context.TODO())
 
 	response := make(chan DockerContainerMetadata, 1)
-	go func() { response <- dg.pullImage(image, authData) }()
+	go func() {
```

Comment:
Should we distinguish the errors generated by the agent from the errors generated by docker, and we only retry for errors from docker? For the errors generated by the agent like `CannotPullECRContainerError` and `CreateEmptyVolumeError` probably won't recover with retry.

Reply:
`CannotPullECRContainerError` is already excluded from retries.  I can extend `CreateEmptyVolumeError` to exclude it from retries as well.  The other errors wrapped by `CannotPullContainerError` are not excluded right now; I only covered ECR because it was easy.